src/core/border_router/routing_manager.cpp

```diff
@@ -3110,6 +3161,172 @@ void RoutingManager::RsSender::HandleTimer(void)
+    {
+        WithdrawPrefix();
+    }
+}
+
+Error RoutingManager::PdPrefixManager::Process(const Ip6::Nd::RouterAdvertMessage &aMessage)
+{
+    Error                        error = kErrorNone;
+    DiscoveredPrefixTable::Entry favoredEntry;
+    bool                         currentPrefixUpdated = false;
+
+    VerifyOrExit(aMessage.IsValid(), error = kErrorParse);
+    favoredEntry.Clear();
+
+    for (const Ip6::Nd::Option &option : aMessage)
+    {
+        DiscoveredPrefixTable::Entry entry;
+
+        if (option.GetType() != Ip6::Nd::Option::Type::kTypePrefixInfo ||
+            !static_cast<const Ip6::Nd::PrefixInfoOption &>(option).IsValid())
+        {
+            continue;
+        }
+
+        entry.SetFrom(static_cast<const Ip6::Nd::PrefixInfoOption &>(option));
+
+        // We should accept ULA prefix since it could be used by the internet infrastructure like NAT64.
+        if (entry.GetPrefix().GetLength() > kOmrPrefixLength || entry.GetPrefix().GetLength() == 0 ||
+            entry.GetPrefix().IsLinkLocal() || entry.GetPrefix().IsMulticast())
+        {
+            LogWarn("PdPrefixManager: Ignore invalid PIO entry %s", entry.GetPrefix().ToString().AsCString());
+            continue;
+        }
+
+        entry.mPrefix.SetLength(kOmrPrefixLength);
+        entry.mPrefix.Tidy();
+
+        // The platform may send another RA message to announce that the current prefix we are using is no longer
+        // preferred or no longer valid.
+        if (entry.GetPrefix() == GetPrefix())
+        {
+            currentPrefixUpdated = true;
+            mPrefix              = entry;
+        }
+
+        if (entry.IsDeprecated())
+        {
+            continue;
+        }
+
+        // Some platforms may delegate us more than one prefixes. We will pick the smallest one. This is a simple rule
+        // to pick the GUA prefix from the RA messages since GUA prefixes (2000::/3) are always smaller than ULA
+        // prefixes (fc00::/7).
+        if (favoredEntry.GetPrefix().GetLength() == 0 || entry.GetPrefix() < favoredEntry.GetPrefix())
+        {
+            favoredEntry = entry;
+        }
+    }
+
+    if (currentPrefixUpdated && mPrefix.IsDeprecated())
+    {
+        LogInfo("PdPrefixManager: Prefix %s is deprecated", mPrefix.GetPrefix().ToString().AsCString());
+        mPrefix.Clear();
+        Get<RoutingManager>().ScheduleRoutingPolicyEvaluation(kImmediately);
+    }
+
+    if (!HasPrefix() || (favoredEntry.GetPrefix().GetLength() != 0 && favoredEntry.GetPrefix() < mPrefix.GetPrefix()))
+    {
+        mPrefix = favoredEntry;
+        Get<RoutingManager>().ScheduleRoutingPolicyEvaluation(kImmediately);
+    }
+
+exit:
+    if (HasPrefix())
+    {
+        mTimer.FireAt(mPrefix.GetStaleTime());
+    }
+    else
+    {
+        mTimer.Stop();
+    }
+
+    return error;
+}
+
+void RoutingManager::PdPrefixManager::ProcessPlatformGeneratedRa(const uint8_t *aRouterAdvert, const uint16_t aLength)
```

Comment:
Nit: Move this before `Process()`? Make it easier as reading the code from top-down?